registry/remote/repository.go

```diff
@@ -944,61 +952,130 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
+		return ocispec.Descriptor{}, fmt.Errorf(
+			"%s %q: invalid response `Content-Type` header; %w",
+			resp.Request.Method,
+			resp.Request.URL,
+			err,
+		)
+	}
+
+	// 2. Validate Size
+	if resp.ContentLength == -1 {
+		return ocispec.Descriptor{}, fmt.Errorf(
+			"%s %q: unknown response Content-Length",
+			resp.Request.Method,
+			resp.Request.URL,
+		)
+	}
+
+	// 3. Validate Client Reference
+	var refDigest digest.Digest
+	if d, err := ref.Digest(); err == nil {
+		refDigest = d
+	}
+
+	// 4. Validate Server Digest (if present)
+	var serverHeaderDigest digest.Digest
+	if serverHeaderDigestErr := resp.Header.Get(dockerContentDigestHeader); serverHeaderDigestErr != "" {
+		if serverHeaderDigest, err = digest.Parse(serverHeaderDigestErr); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf(
+				"%s %q: invalid response header value `%s`: `%s`; %w",
+				resp.Request.Method,
+				resp.Request.URL,
+				dockerContentDigestHeader,
+				serverHeaderDigestErr,
+				err,
+			)
+		}
 	}
 
-	size := resp.ContentLength
-	if size == -1 {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: unknown response Content-Length", resp.Request.Method, resp.Request.URL)
-	}
+	/* 5. Now, look for specific error conditions; see truth table in method docstring */
+	var contentDigest digest.Digest
 
-	// validate digest if ref is a digest
-	if refDigest, err := ref.Digest(); err == nil {
-		if err = verifyContentDigest(resp, refDigest); err != nil {
-			return ocispec.Descriptor{}, err
+	var calculatedDigest digest.Digest
+	if len(serverHeaderDigest) == 0 {
+		if httpMethod == http.MethodHead {
+			if len(refDigest) == 0 {
+				// HEAD without server `Docker-Content-Digest` header is an
+				// immediate fail; HEAD
+				return ocispec.Descriptor{}, fmt.Errorf(
+					"HTTP %s request missing required header `%s`",
+					httpMethod, dockerContentDigestHeader,
+				)
+			}
+		} else {
+			// GET without server `Docker-Content-Digest` header forces the
+			// expensive calculation
+			if calculatedDigest, err = calculateDigestFromResponse(resp, s.repo.MaxMetadataBytes); err != nil {
+				return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
+			}
+			contentDigest = calculatedDigest
 		}
-		return ocispec.Descriptor{
-			MediaType: mediaType,
-			Digest:    refDigest,
-			Size:      size,
-		}, nil
-	}
-
-	digestStr := resp.Header.Get("Docker-Content-Digest")
-	if digestStr == "" {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: empty response Docker-Content-Digest", resp.Request.Method, resp.Request.URL)
+	} else {
+		contentDigest = serverHeaderDigest
 	}
 
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+	if len(refDigest) > 0 && len(contentDigest) > 0 && refDigest != contentDigest {
```

Comment:
We don't need to and should not check `len(contentDigest) > 0`. If `len(refDigest) > 0` is true but `len(contentDigest) == 0`, then we will always have `refDigest != contentDigest`.

Again, if `len(contentDigest) == 0`, the entire Boolean expression is false. Therefore, it passes the check, and we will return a descriptor with an empty digest, which is problematic.

Reply:
Now since we're trusting the client, and setting `contentDigest = refDigest` as per https://github.com/oras-project/oras-go/pull/237#discussion_r954467389, yes I can again drop this change (this was just added in the last push).

Done.